### DISCUSSION TITLE
What is the purpose of node_modules inside virtual store?

### DISCUSSION THREAD

**Discussion by miqdad-f** [pnpm/pnpm#4224] [Q&A] [answered]
Hi, sorry for a newbie question here. Just curious, after digging layout of `node_modules` generated by pnpm, I noticed there is additional `node_modules` inside `node_modules/.pnpm`. I have read some background motivation docs of pnpm but I could not find any explanation about this. So what is the objective of `node_modules/.pnpm/node_modules`? 

Any explanation would be appreciated, thank you 🙏🏼

**Comment by zkochan** [comment, marked as answer]
There are many broken packages in the npm ecosystem that try to require packages that are not in their dependencies. In order to make them work with pnpm, we hoist all deps to `node_modules/.pnpm/node_modules`. You can control this with the `hoist` and `hoist-pattern` options.

**Reply by rrosenshain-sc** [reply]
Is there a reason why which package gets hoisted in non-deterministic?
